### PR TITLE
Core: Pre-bundle `@storybook/theming` to avoid emotion11 conflicts

### PR DESCRIPTION
**Pull request by ndelangen**
Issue: #16716 #13145 #13114

introduce pre-bundle script as an alternative to prepare script.

pre-bundling means we'll:

- create a runtime bundle of the entire package.
  we actually create the bundle 3 times (CJS, ESM, modern)
- create a collected type-definitions file

We can specify our source-code entry-point in our package's package.json file:
```
  "bundlerEntrypoint": "./src/index.ts",
```

And we can specify which dependencies we'd like to NOT be bundled (will apply to both the runtime bundles and type-definitions) by adding those modules to the `dependencies` and `peerDependencies` properties in `package.json`

To use the prebundle script instead of the prepare script:
```
  "scripts": {
    "prepare": "ts-node ../../scripts/prebundle.ts"
  },
```
You may need to install `ts-node` in the package for this to work.

TODO:
- [x] fix bootstrap
- [x] fix linting
- [x] fix tests
- [x] investigate if I can generate `dist/ts3.4` again.
- [x] implement some sort of watch mode for `yarn dev`

### CODE REVIEW

File: lib/theming/package.json [hunk 1]
```diff
@@ -60,6 +63,13 @@
...
-  "sbmodern": "dist/modern/index.js"
+  "bundlerEntrypoint": "./src/index.ts",
+  "gitHead": "ce3592fc323ede3d33939b532ff7400cfdc1efa1",
+  "sbmodern": "dist/modern/index.js",
+  "unbundledDependencies": [
+    "react",
+    "react-dom",
+    "core-js",
+    "@storybook/client-logger"
+  ]
```
Comment (ndelangen) [inline lib/theming/package.json:74]: should we derive this from `dependencies` and `peerDependencies` ?

File: lib/theming/src/index.ts [hunk 2]
```diff
@@ -1,3 +1,6 @@
+// eslint-disable-next-line @typescript-eslint/triple-slash-reference
+/// <reference path="./typings.d.ts" />
+
```
Comment (ndelangen) [inline lib/theming/src/index.ts:3]: Is this still needed?

File: lib/theming/tsconfig.json [hunk 3]
```diff
@@ -3,15 +3,20 @@
...
     "src/**/__testfixtures__/**"
-  ]
+  ],
+  "ts-node": {
+    "transpileOnly": true,
+    "files": true,
+    "compilerOptions": {
+      "types": ["node"]
+    }
+  }
```
Comment (ndelangen) [inline lib/theming/tsconfig.json:21]: should be inherited from the root `tsconfig.json`?

File: scripts/prebundle.ts [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+  await buildESM(options);
+  await buildCJS(options);
+  await buildModern(options);
+
+  const [out] = await generateDtsBundle([
+    { filePath: input, output: { inlineDeclareGlobals: true, sortNodes: true, noBanner: true } },
+  ]);
+
+  const bundledDTSfile = path.join(cwd, 'dist/ts-tmp/index.d.ts');
```
Comment (ndelangen) [inline scripts/prebundle.ts:137]: We should put this in a different place, so that we do not publish to npm

File: lib/theming/src/index.ts (hunk 2 above)
Comment (ndelangen) [reply lib/theming/src/index.ts:3]: YES

File: .eslintrc.js [hunk 5]
```diff
@@ -72,5 +72,12 @@ module.exports = {
         'react/no-unknown-property': 'off', // Need to deactivate otherwise eslint replaces some unknown properties with React ones
       },
     },
+    {
+      // this package uses pre-bundling, dependencies will be bundled, and will be in devDepenencies
+      files: ['**/lib/theming/**/*'],
+      rules: {
+        'import/no-extraneous-dependencies': 'off',
```
Comment (tmeasday) [inline .eslintrc.js:79]: Can we check they are at least in `devDependencies` though?

File: lib/core-server/src/__snapshots__/vue-3-cli_preview-dev-posix [hunk 6]
```diff
@@ -466,8 +466,6 @@ Object {
         "options": Object {
           "docgenOptions": Object {
             "alias": Object {
-              "@emotion/core": "NODE_MODULES/@emotion/core",
-              "@emotion/styled": "NODE_MODULES/@emotion/styled",
               "@storybook/addons": "ROOT/lib/addons",
```
Comment (tmeasday) [inline lib/core-server/src/__snapshots__/vue-3-cli_preview-dev-posix:469]: Love to see it

File: scripts/dts-localize.ts [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+
+  for (const sourceFile of sourceFiles) {
+    // skip lib files from the compiler
+    if (program.isSourceFileDefaultLibrary(sourceFile)) {
+      continue;
+    }
+
+    // here you can implement any logic regarding how your new files should be placed
+    // as for this POC I used rough solution to use the same structure but I believe it should be improved
+    filesRemapping.set(sourceFile.fileName, getNewPath(cwd, sourceFile.fileName));
```
Comment (tmeasday) [inline scripts/dts-localize.ts:93]: Is it necessary to do the mapping for all source files up front? As you walked me through it it seemed like instead we could do it on demand. I'm not sure how many files we are talking about or how expensive it is, perhaps that's a premature optimisation.

File: scripts/dts-localize.ts [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+      const filename = parts.join('node_modules/').split('/').join('-');
+      newPath = path.join(outputPath, '_modules', filename);
+    } else if (relative.includes('dist/ts-tmp/')) {
+      const [, ...parts] = relative.split('dist/ts-tmp/');
+      const filename = parts.join('').split('/').join('-');
+      newPath = path.join(outputPath, filename);
+    } else {
+      const filename = relative.split('/').join('-');
+      newPath = path.join(outputPath, filename);
+    }
```
Comment (tmeasday) [inline scripts/dts-localize.ts:117]: I think you were going to add a comment explaining this

File: scripts/dts-localize.ts [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+    return newPath;
+  }
+
+  function wasReplaceAlready(fileName: string, target: string) {
```
Comment (tmeasday) [inline scripts/dts-localize.ts:121]:
```suggestion
  function wasReplacedAlready(fileName: string, target: string) {
```

File: scripts/dts-localize.ts [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+      .slice(1)
+      .replace('.d.ts', '')
+      .replace('.ts', '');
+
+    replaceRemapping.set(currentSourceFile, [
+      ...(replaceRemapping.get(currentSourceFile) || []),
+      targetRelativeToSource2,
+    ]);
+
+    console.log('renamed', target);
```
Comment (tmeasday) [inline scripts/dts-localize.ts:145]: Going to drop these logs?

File: scripts/dts-localize.ts [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+      ).fileName;
+
+      const replacementPath = getReplacementPath(currentSourceFile, referencedSourceFile, target);
+
+      // @ts-ignore
+      node.argument = ts.createStringLiteral(replacementPath);
+      // node.argument = ts.factory.createStringLiteral(replacementPath); // TS4
+
+      return true;
+    }
```
Comment (tmeasday) [inline scripts/dts-localize.ts:217]: This seems very duplicative with the block above

File: scripts/prebundle.ts [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+
+  const options = {
+    cwd,
+    externals,
+    input,
+  };
+
+  await buildESM(options);
+  await buildCJS(options);
+  await buildModern(options);
```
Comment (tmeasday) [inline scripts/prebundle.ts:132]: Could this happen in parallel? Would that be a bad idea?

**Review by tmeasday — COMMENTED** [review]
Love it @ndelangen

File: scripts/dts-localize.ts [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+    filesRemapping.set(sourceFile.fileName, getNewPath(cwd, sourceFile.fileName));
+  }
+
+  entrySourceFiles.forEach((file) => {
+    const sourceFile = sourceFiles.find((f) => f.fileName === file);
+
+    actOnSourceFile(sourceFile);
+  });
+
+  function getNewPath(basePath: string, filePath: string) {
```
Comment (tooppaaa) [inline scripts/dts-localize.ts:102]: Worth to give a try on windows !

File: scripts/prebundle.ts (hunk 12 above)
Comment (ndelangen) [reply scripts/prebundle.ts:132]: perhaps I can use a single rollup run to generate all 3..

```
output: [
    {
      format: 'esm',
      dir: 'output',
    },
    {
      format: 'system',
      dir: 'output/legacy',
    },
  ],
  ```

File: scripts/dts-localize.ts (hunk 11 above)
Comment (ndelangen) [reply scripts/dts-localize.ts:217]: It is, but the way the target is read, and how to replacementPath is changed are different. How would you suggest I make this code let duplicative without adding more misdirection?

It'd require at least 3 extra if-else blocks, which might be just as hard to understand?

File: scripts/dts-localize.ts (hunk 8 above)
Comment (ndelangen) [reply scripts/dts-localize.ts:120]: I did!

**Review by azooz-12 — APPROVED** [review]

**Comment by ndelangen** [comment]
I'm trying to make this new way of 'preparing' work for out existing "dev" script, and found out we have a few ways to do "dev".

I'm simplifying then and removing some of the unused code, and cleaning up a bit.

**Comment by ndelangen** [comment]
OK, I think it should work now, and quite a few changes I had to do...
I am going to split it into 2 separate PRs.

**Pull request merged**